---
title: "Luigi's COVID-19 Data"
output: html_notebook
---

Version: 3.0  
Created: 9/6/2023  
Last update: 9/22/2023

Analysis datasets and copy of the Rmd are in Rutgers box:  
https://rutgers.app.box.com/folder/227391194192  

GitHub Repo:  
https://github.com/sargdavid/luigi_covid19  
  
# Setup
```{r setup}
require(data.table)
require(ggplot2)
require(table1)
require(DESeq2)
```

Javier, 9/15/2023:  
1. Build survival model using LOS as dates. Assuming all death are in-hospital, end of LOS.  


# Data
## Clinical variables
```{r}
dt_clin <- fread("Data/COVID study Master sheet w ELISA data July 2023-2.csv")
```

## Remove Patient 7247980
Patient 7247980 was marked “COVID-Positive” instead of 1-3 in Inpatient=1, ER=2, OP=3, and had most data missing. Removed from the analysis. Luigi agreed.
```{r}
dt_clin <- dt_clin[`Real ID` != 7247980, ]

# # Remove 1 patient marked 'COVID-Positive': SAME patient!
# dt_clin[`Inpatient=1, ER=2, OP=3` == "COVID-Positive",]
# dt_clin <- dt_clin[`Inpatient=1, ER=2, OP=3` != "COVID-Positive", ]
```

### Format clinical variables
```{r}
dt_clin$`Real ID` <- as.character(dt_clin$`Real ID`)

dt_clin$`Whole Blood` <- factor(dt_clin$`Whole Blood`,
                            levels = unique(dt_clin$`Whole Blood`))

dt_clin$Plasma <- factor(dt_clin$Plasma)

# COVID = 0/1
dt_clin$`Study Group` <- factor(dt_clin$`Study Group`,
                            levels = 0:1,
                            labels = c("No COVID",
                                       "COVID"))

dt_clin$`Inpatient=1, ER=2, OP=3` <- factor(dt_clin$`Inpatient=1, ER=2, OP=3`,
                                        levels = 1:3,
                                        labels = c("Inpatient",
                                                   "ER",
                                                   "OP"))

dt_clin$Sex <- factor(dt_clin$Sex,
                  levels = 0:1,
                  labels = c("Male",
                             "Female"))

# Keys from Lu
# NOTE: =4 is not in the Keys!
dt_clin$Race <- factor(dt_clin$Race,
                   levels = 0:4,
                   labels = c("White Non-Hispanic",
                              "Black",
                              "Indian/Asian",
                              "Other",
                              "Unknown"))

dt_clin$`If BMI ?30 Obesity Class (I,II,III)` <- factor(dt_clin$`If BMI ?30 Obesity Class (I,II,III)`,
                                                    levels = c("I",
                                                               "II",
                                                               "III"))

# !!!GET LABELS!!!
dt_clin$Vaccinated <- factor(dt_clin$Vaccinated,
                         levels = c(0, 1, 9))

dt_clin$`Dose #1` <- factor(dt_clin$`Dose #1`,
                        levels = c(0, 1, 9))

dt_clin$`Dose #2` <- factor(dt_clin$`Dose #2`,
                        levels = c(0, 1, 9))

dt_clin$Booster <- factor(dt_clin$Booster,
                      levels = c(0, 1, 9))

dt_clin$`Critical Care` <- factor(dt_clin$`Critical Care`,
                              levels = 0:1)

# LOS = length of stay in the hospital
dt_clin$Still_In_Hospital_08182023 <- is.na(as.numeric(dt_clin$LOS))
dt_clin$LOS <- as.numeric(sapply(strsplit(x = dt_clin$LOS, 
                                      split = " "), 
                             function(x) x[1]))

# Death
dt_clin$Expired <- factor(dt_clin$Expired,
                      levels = 0:1)

dt_clin$Wave <- factor(dt_clin$Wave,
                   levels = c(1:4))

# Luigi's keys
dt_clin$`Smoking History (0,1,2)` <- factor(dt_clin$`Smoking History (0,1,2)`,
                                        levels = 0:2,
                                        labels = c("Never",
                                                   "Former",
                                                   "Current"))

# !!! GET LABELS!!!
dt_clin$Age_Category <- factor(dt_clin$Age_Category,
                               levels = 0:4)

# Co-diagnoses/pre-existing conditions
dt_clin$MI <- factor(dt_clin$MI)
dt_clin$HF <- factor(dt_clin$HF)
dt_clin$PVD <- factor(dt_clin$PVD)
dt_clin$CVD <- factor(dt_clin$CVD)
dt_clin$Dementia <- factor(dt_clin$Dementia)
dt_clin$COPD <- factor(dt_clin$COPD)
dt_clin$`Rheum/ CONNECTIVE TISSUE DISEASE` <- factor(dt_clin$`Rheum/ CONNECTIVE TISSUE DISEASE`)
dt_clin$PUD <- factor(dt_clin$PUD)
dt_clin$mliverd <- factor(dt_clin$mliverd)
dt_clin$DiabNC <- factor(dt_clin$DiabNC)
dt_clin$DiabC <- factor(dt_clin$DiabNC)
dt_clin$AnyDM <- factor(dt_clin$AnyDM)
dt_clin$Hemoplegia <- factor(dt_clin$Hemoplegia)
dt_clin$RenalD <- factor(dt_clin$RenalD)
dt_clin$Cancer <- factor(dt_clin$Cancer)
dt_clin$msliver <- factor(dt_clin$msliver)
dt_clin$MetaCancer <- factor(dt_clin$MetaCancer)
dt_clin$AIDS <- factor(dt_clin$AIDS)
dt_clin$HTN <- factor(dt_clin$HTN)
dt_clin$Obese <- factor(dt_clin$Obese)

# Medications
dt_clin$tocilizumab <- factor(dt_clin$tocilizumab)
dt_clin$remdesivir <- factor(dt_clin$remdesivir)
dt_clin$dexamethasone <- factor(dt_clin$dexamethasone)
dt_clin$methylprednisolone <- factor(dt_clin$methylprednisolone)
dt_clin$prednisone <- factor(dt_clin$prednisone)
dt_clin$hydrocortisone <- factor(dt_clin$hydrocortisone)
dt_clin$azithromycin <- factor(dt_clin$azithromycin)
dt_clin$hydroxychloroquine <- factor(dt_clin$hydroxychloroquine)
dt_clin$regeneron <- factor(dt_clin$regeneron)
dt_clin$ACEis <- factor(dt_clin$ACEis)
dt_clin$ARBs <- factor(dt_clin$ARBs)
dt_clin$ARNIs <- factor(dt_clin$ARNIs)
dt_clin$Insulin <- factor(dt_clin$Insulin)
dt_clin$Metformin <- factor(dt_clin$Metformin)
dt_clin$Glimepiride <- factor(dt_clin$Glimepiride)
dt_clin$Glipizide <- factor(dt_clin$Glipizide)
dt_clin$Sitagliptin <- factor(dt_clin$Sitagliptin)
dt_clin$`Full Dose Anticoagulation` <- factor(dt_clin$`Full Dose Anticoagulation`)
dt_clin$`Prophylactic Anticoagulation` <- factor(dt_clin$`Prophylactic Anticoagulation`)

# Lab work
## NOTE: set correct values for LOQ/LOD! Converted to NAs for now
dt_clin$HbA1c <- as.numeric(dt_clin$HbA1c)
dt_clin$Vit_D_level <- as.numeric(dt_clin$Vit_D_level)
dt_clin$ALT <- as.numeric(dt_clin$ALT)
dt_clin$AST <- as.numeric(dt_clin$AST)
dt_clin$`hs-CRP` <- as.numeric(dt_clin$`hs-CRP`)
dt_clin$CRP <- as.numeric(dt_clin$CRP)
dt_clin$Fibrinogen <- as.numeric(dt_clin$Fibrinogen)
dt_clin$`D-Dimer` <- as.numeric(dt_clin$`D-Dimer`)
dt_clin$ESR <- as.numeric(dt_clin$ESR)
dt_clin$Ferritin <- as.numeric(dt_clin$Ferritin)
dt_clin$LDH <- as.numeric(dt_clin$LDH)
dt_clin$`Lactic Acid` <- as.numeric(dt_clin$`Lactic Acid`)
dt_clin$PCT <- as.numeric(dt_clin$PCT)
dt_clin$`Glucose Lvl` <- as.numeric(dt_clin$`Glucose Lvl`)
dt_clin$Screat <- as.numeric(dt_clin$Screat)

# Diagnoses
dt_clin$Hyperlipidemia <- factor(dt_clin$Hyperlipidemia)
dt_clin$`# Metabolic Syndrome Criteria` <- factor(dt_clin$`# Metabolic Syndrome Criteria`)

# NA should be non-COVID patients - CHECK
dt_clin$`WHO-OSCI D1` <- factor(dt_clin$`WHO-OSCI D1`,
                            levels = 0:7)
dt_clin$`WHO-OSCI D3` <- factor(dt_clin$`WHO-OSCI D3`,
                            levels = 0:7)
dt_clin$`WHO-OSCI D7` <- factor(dt_clin$`WHO-OSCI D7`,
                            levels = 0:7)
dt_clin$`WHO-OSCI D14` <- factor(dt_clin$`WHO-OSCI D14`,
                             levels = 0:7)
dt_clin$`WHO-OSCI D21` <- factor(dt_clin$`WHO-OSCI D21`,
                             levels = 0:7)
dt_clin$`WHO-OSCI D28` <- factor(dt_clin$`WHO-OSCI D28`,
                             levels = 0:7)
dt_clin$`WHO-OSCI DISCHARGE` <- factor(dt_clin$`WHO-OSCI DISCHARGE`)

# Protein columns have no data - need to merge with another file. Remove last 21 columns.
dt_clin <- dt_clin[, Fake_ID:`WHO-OSCI DISCHARGE`]
```

```{r}
summary(dt_clin)
```

## Proteins
```{r}
dt_elisa <- fread("Data/COVID study Master sheet w ELISA data July 2023-3.csv")
```

### Check matching IDs
```{r}
# There are 4 duplicated IDs
dt_elisa[`Sample ID` %in% `Sample ID`[which(duplicated(`Sample ID`))]]

# FIX THIS! 
# For now, keep 1 of 2 samples
dt_elisa <- dt_elisa[!(`Relabeled ID` %in% c("P122",
                                             "P38",
                                             "P2",
                                             "P120")), ]

# NOTE: checked other types of ID matches - no good
# dt_elisa$`Sample ID`[!(dt_elisa$`Sample ID` %in% dt_clin$Fake_ID)]
# dt_elisa$`Relabeled ID`[!(dt_elisa$`Relabeled ID` %in% dt_clin$`Real ID`)]
# dt_elisa$`Relabeled ID`[!(dt_elisa$`Relabeled ID` %in% dt_clin$Fake_ID)]

dt_elisa$`Sample ID`[!(dt_elisa$`Sample ID` %in% dt_clin$`Real ID`)]

length(unique(dt_elisa$`Sample ID`))
```

### Merge matching Real IDs
```{r}
dt_elisa$`Relabeled ID` <- NULL
names(dt_elisa)[1] <- "Real ID"

dt_clin$Fake_ID <- NULL

dt_clin <- merge(dt_clin,
                 dt_elisa,
                 by = "Real ID",
                 all.x = TRUE)
```

## Gene expressions
NOTE: data contains both, coding and noncoding genes.  
  
### First RNA-seq file
```{r}
dt2 <- fread("Data/gene_count.xls")

rnaseq_meta1 <- data.table(sample_names = colnames(dt2)[2:58],
                           rnaseq_cohort = sapply(strsplit(x = colnames(dt2)[2:58], 
                                                           split = "_"), 
                                                  function(x) x[1]),
                           `Real ID` = sapply(strsplit(x = colnames(dt2)[2:58], 
                                                       split = "_"), 
                                              function(x) x[2]))
rnaseq_meta1
```

### Second RNA-seq file
```{r}
dt3 <- fread("Data/gene_count_Additional.txt")

rnaseq_meta2 <- data.table(sample_names = names(dt3)[2:46],
                           rnaseq_cohort = "",
                           `Real ID` = gsub(pattern = "LB", 
                                            x = names(dt3)[2:46],
                                            replacement = ""))
rnaseq_meta2
```

### Merge RNA-seq meta files
```{r}
rnaseq_meta <- rbindlist(list(rnaseq_meta1,
                              rnaseq_meta2))
length(unique(rnaseq_meta$`Real ID`))
# Same as number of rows, hence, all unique sample IDs

rnaseq_meta$rnaseq_cohort <- NULL
```

### Merge RNA-seq files
First 10 columns are gene info, the other 102 columns are the samples.
```{r}
dt_rna <- merge(dt2,
                dt3,
                by = c("gene_name",
                       "gene_id",
                       "gene_chr",
                       "gene_start",
                       "gene_end",
                       "gene_strand",
                       "gene_length",
                       "gene_biotype",
                       "gene_description",
                       "tf_family"))
```

### IDs that are in RNA-seq but not in Clinical
10 IDs  
```{r}
id_no_clin <- rnaseq_meta$`Real ID`[!(rnaseq_meta$`Real ID` %in% dt_clin$`Real ID`)]
id_no_clin
```

### IDs that are in Clinical but not in RNA-seq
91 IDs  
```{r}
id_no_rna <- dt_clin$`Real ID`[!(dt_clin$`Real ID` %in% rnaseq_meta$`Real ID`)]
id_no_rna 
```

## Save data
```{r}
save(dt_clin,
     file = "Data/dt_clin.RData")

save(dt_rna,
     file = "Data/dt_rna.RData")

save(rnaseq_meta,
     file = "Data/rnaseq_meta.RData")

write.csv(dt_clin,
          file = "Data/luigi_covid19_clinical_data.csv",
          row.names = FALSE)

write.csv(dt_rna,
          file = "Data/luigi_covid19_nra_seq_data.csv",
          row.names = FALSE)
```

## Clean
```{r}
rm(list = ls())

# rm(list = ls()[!ls() %in% c("dt_clin",
#                             "dt_rna",
#                             "rnaseq_meta")])

gc()
```

# Reload data
```{r}
load("Data/dt_clin.RData")
load("Data/dt_rna.RData")
load("Data/rnaseq_meta.RData")
```

# Analysis
## Table 1
```{r}
t1 <- table1( ~.| `Study Group` + AnyDM,
              data = dt_clin[, c(-1)])

t1
```

```{r}
write.table (t1 , 
             file = "Output/table1.csv",
             col.names = TRUE, 
             row.names = FALSE,
             append = TRUE,
             sep = ',')
```

## Death
### Death vs. COVID
```{r}
addmargins(table(Death = dt_clin$Expired,
                 COVID = dt_clin$`Study Group`))
```

### Death vs DM,  COVID only 
```{r}
covid <- droplevels(dt_clin[`Study Group` == "COVID", ])

chisq.test(covid$Expired,
           covid$AnyDM)

addmargins(table(Death = covid$Expired,
                 AnyDM = covid$AnyDM))
```

### Death vs Obese, COVID only 
```{r}
chisq.test(dt_clin$Expired[dt_clin$`Study Group` == "COVID"],
           dt_clin$Obese[dt_clin$`Study Group` == "COVID"])

addmargins(table(Death = dt_clin$Expired[dt_clin$`Study Group` == "COVID"],
                 Obese = dt_clin$Obese[dt_clin$`Study Group` == "COVID"]))
```

### Death vs BMI, COVID only 
```{r}
t.test(dt_clin$BMI[dt_clin$`Study Group` == "COVID"] ~ dt_clin$Expired[dt_clin$`Study Group` == "COVID"])

ggplot(dt_clin[`Study Group` == "COVID", ],
       aes(x = Expired,
           y = BMI)) +
  geom_boxplot(outlier.shape = "") +
  geom_point(aes(group = `Real ID`),
             shape = 21,
             size = 2,
             position = position_dodge(0.3)) +
  ggtitle("COVID Patients Only") +
  theme_bw()
```

### Death vs Critical Care
Significant
```{r}
addmargins(table(Death = covid$Expired,
                 CrCare = covid$`Critical Care`))

m3 <- glm(Expired ~ `Critical Care`,
          family = binomial,
          data = covid)
s3 <- summary(m3)
s3$coefficients
s3
exp(s3$coefficients[, 1])

chisq.test(covid$Expired,
           covid$`Critical Care`)
```

#### CONTINUE HERE!!! 9/22/2023

### Death vs. COVID | Diabetes
No significant difference.  
Model with interaction is worse.  
  
```{r}
dtm2 <- dt_clin[!is.na(Expired) &
              `Study Group` != "Unknown" &
              AnyDM != "Diabetes Unknown", ]


m2 <- glm(Expired ~ `Study Group` +
            AnyDM,
          family = binomial,
          data = droplevels(dtm2))
summary(m2)
```



### Death vs COVID, adjusted for critical care
Both significant in additive model.  
No significance in interaction model.
```{r}
dtm4 <- dt_clin[!is.na(Expired) &
              AnyDM != "Diabetes Unknown" &
              `Study Group` != "Unknown" &
              !is.na(`Critical Care`), ]

m4 <- glm(Expired ~ `Study Group` + `Critical Care`,
          family = binomial,
          data = droplevels(dtm4))
summary(m4)

m41 <- glm(Expired ~ `Study Group`*`Critical Care`,
          family = binomial,
          data = droplevels(dtm4))
summary(m41)
```

### Death vs COVID, adjusted for DM
No significance
```{r}
dtm5 <- dt_clin[!is.na(Expired) &
              `Study Group` != "Unknown" &
              AnyDM != "Diabetes Unknown", ]

m5 <- glm(Expired ~ `Study Group` +
            AnyDM,
          family = binomial,
          data = droplevels(dtm5))
summary(m5)
```

### Death vs COVID, adjusted 
```{r}
dtm6 <- dt_clin[!is.na(Expired) &
              `Study Group` != "Unknown" &
              AnyDM != "Diabetes Unknown" &
              !is.na(Age_Category), ]

m6 <- glm(Expired ~ `Study Group` +
            AnyDM +
            Age_Category +
            Sex +
            Race +
            BMI,
          family = binomial,
          data = droplevels(dtm6))
summary(m6)
```

### New variable: DM+Obesity
```{r}
chisq.test(dt_clin$AnyDM,
           dt_clin$Obese)
```

## ICU
### ICU LOS
```{r}
summary(dt_clin$LOS)
hist(dt_clin$LOS)
hist(log2(dt_clin$LOS + 1))

m2 <- lm(log2(LOS + 1) ~ AnyDM +
           Age +
           Sex +
           Race +
           BMI +
           `Critical Care`,
         family = binomial,
         data = droplevels(dt_clin[!is.na(Expired) &
                                 AnyDM != "Diabetes Unknown", ]))
summary(m2)
```

### Critical care
```{r}
m3 <- glm(`Critical Care` ~ AnyDM +
            Age +
            Sex +
            Race +
            BMI,
          family = binomial,
          data = droplevels(dt_clin[!is.na(Expired) &
                                  AnyDM != "Diabetes Unknown", ]))
summary(m3)
```

## WHO-OSCI
### WHO-OSCI over time
```{r}
tmp <- dt_clin[, c(1, 5, 37, 82:87)]
tmp <- melt.data.table(tmp,
                       id.vars = c(1:3),
                       variable.name = "Day",
                       value.name = "WHO_OSCI")
tmp$Day <- factor(tmp$Day,
                       levels = unique(tmp$Day))

tmp <- tmp[!is.na(tmp$WHO_OSCI)]
ggplot(tmp,
       aes(x = Day,
           y = WHO_OSCI,
           group = `Real ID`,
           color = `Real ID`)) +
  facet_wrap(~ AnyDM) +
  geom_line(position = position_dodge(0.3)) +
  theme_bw() +
  theme(legend.position = "none") +
  theme(axis.text.x = element_text(angle = 45,
                                   hjust = 1))
```

## RNA-seq
### DESeq2
```{r}
rnaseq_meta$rnaseq_cohort <- factor(rnaseq_meta$rnaseq_cohort,
                                    levels = c("ctl",
                                               "cov"))
tmp <- data.frame(dt2[, cov_464811:ctl_784880])
rownames(tmp) <- dt2$gene_id

dds <- DESeqDataSetFromMatrix(countData = dt2[, cov_464811:ctl_784880], 
                              colData = rnaseq_meta,
                              ~ rnaseq_cohort)
```

```{r}
# dss <- estimateSizeFactors(dds, 
#                            type = "iterate")

deseq <- DESeq(dds)
```

```{r}
res <- results(deseq)
res[1:10, ]

out <- data.table(gene_id = rownames(res),
                  res)
```

# Session
```{r}
sessionInfo()
```