---
title: "Data Analysis: Genetic and inflammatory signatures associated with worse prognosis in hospitalized patients with severe SARS-CoV-2 infection and diabetes."
author: Davit Sargsyan
output:
  html_notebook:
    highlight: tango
    toc: true
    toc_float: true
    toc_depth: 6
    number_sections: yes
    code_folding: hide
---

Created: 9/6/2023\
Last update: 3/15/2024

# Background

Analysis datasets and copy of the Rmd are in Rutgers box:\
<https://rutgers.app.box.com/folder/227391194192>

GitHub Repo:\
<https://github.com/sargdavid/luigi_covid19>

**Meeting 3/14/2024:**

1.  Lu: cytokine storm can more likely happen to the ICU/sever patients. Therefore, stratify by severity etc.
2.  My ELISA set is still missing ACE2 and DPP4. I need to check the data to see why these columns got deleted.
3.  The 21 proteins included in the attached were measured in pg/mL but the missing ones in ug/mL.
4.  `twopartm::AME` did not converge with the bootstrap option for the 95%CI. Therefore, I used the delta method. Need to check the convergence issue tomorrow with Javier and Yajie.

**ToDo-s:**

1.  Explore why DM prevents INFa and IL-10 increase in COVID patients (i.e., why concentrations of these proteins is high COVID/No DM but not in COVID/DM).
2.  What are the profiles of pritein-non-producing patients?
3.  Use Log2 in ELISA so the differences are 2-fold changes.
4.  On forest plots, show ratios, not log-ratios.
5.  Transform the data using quantile or Fisher-Yates normalization (`DNAMR::cnormalize`)
6.  Apply tow-part model to normalized gene expressions.
7.  Reactom Pathway Analysis (`ReactomPA`) repeated on these sets of genes
8.  Factor analysis of gene expressions. Compare the factors with Gene Ontology, Reactom Pathways, KEGG pathways etc. Include Module Analysis (see [WGCNA: Weighted Correlation Network Analysis](https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-9-559) by Steve Horvath. We used it in [the lncRNA paper](https://insight.jci.org/articles/view/168988?utm_source=TrendMD&utm_medium=cpc&utm_campaign=JCI_Insight_TrendMD_0). Can be more applicable for Paper 2 on lncRNAs.)

**Feedback from stats AI meeting 3/8/2024:**

1.  Use [R package MOFA](https://www.bioconductor.org/packages/release/bioc/html/MOFA2.html) (Bernard)
2.  Check R package [Diablo](http://mixomics.org/mixdiablo/) (part of [mixOmix](https://bioconductor.org/packages/release/bioc/html/mixOmics.html)?) and [COMBI](https://www.bioconductor.org/packages/release/bioc/html/combi.html).
3.  Causal representative learning and causal graphs/networks for coding/non-coding RNA interactions (Mingshi).
4.  Color modules (Caesar, Stefan & John).
5.  Factor analysis using `psych::fa` function (instead of `stats::factanal`).
6.  Gene ontology to connect factors to GOs (Javier) 7. Ask Lu if I can share Table 1 with Michael Katehakis

**Meeting 2/1/2024:**

1.  RNA vs. ELISA - need to know coding genes
2.  List of genes from Lu
3.  IFI27: what cyto/ cemokine is associated with it (from ELISA panel)? Cytokines should regulate IFI27
4.  BAMBI vs ACE2 (new ELISA)
5.  Find specific proteins (ELISA) that result in the expression of another gene
6.  Pathway analysis with `ReactomPA` (DONE)
7.  Focus on DM
8.  Factor analysis
9.  Anti-IL6 was used in pandemic. Look at INFg, IL1, IL6 and TNFa

**Javier, 9/15/2023:**

1.  Build survival model using LOS as dates. Assuming all death are in-hospital, end of LOS.

**Meeting 9/19/23:**

1.  Was everybody admitted with COVID-19 symptoms? In Marshall's email from 11/14/23 8:01AM he said "not sure [admission DX] will be necessary since admitting DX will be COVID". Does this mean that controls where also admitted with COVID diagnoses? How?
2.  Do survival for in-patients only using LOS.
3.  Look at LOS of dead patients.
4.  Do DM patients have higher WHO OSCI score at admission?
5.  COVID+DM = more likely to go to ICU vs COVID+noDM? 6. Outcomes: death, LOS, ventilation, ICU.

**Meeting 9/7/23:**

1.  Look at inflammation signature w/wo DM, stratified by COVID.
2.  In COVID, coagulation cascade pathway was affected. See if we can find it in clin. data, ELIZA or RNA-seq. Should coagulation be considered in COVID patients' treatment?
3.  Coding/non-coding gene co-expression.
4.  Activation of inflammatory pathways because we have transcriptome, not just genome.
5.  Neanderthals' genes (see Lu's email).
6.  Important factors: Critical care, WHO OSCI, LOS, ACE inhibitors, age\>55, HTN, DM.

# Setup

```{r setup}
options(scipen = 999)

# if (!require("BiocManager",
#              quietly = TRUE))
# install.packages("BiocManager")
# BiocManager::install("ReactomePA")

require(data.table)
require(ggplot2)
require(table1)
require(twopartm)
require(psych)
require(psychTools)
require(GPArotation)

require(DESeq2)
require(ReactomePA)

require(tidyverse)
require(tidytlg)
require(gridExtra)
require(nnet)

# Not Used:
# install.packages("zlibbioc_1.48.0.tar.gz",
#                  repos = NULL,
#                  type = "source")

# if (!require("BiocManager",
#              quietly = TRUE))
# BiocManager::install("MOFA2")
#

# require(NBZIMM)::lme.zig(fixed = logTPM ~ ibd_infl,
#                          random = ~ 1| Batch,
#                          zi_fixed = ~ ibd_infl,
#                          data = dt,
#                          verbose = FALSE)
# BiocManager::install("clusterProfiler")

# require(clusterProfiler)
# require(gt)
# require(rtables)
# require(zlibbioc)
# require(MOFA2)
```

```{r}
# rows_empty <- function(mat) {
#   unlist(apply(mat, 1, function(x) !any(nzchar(x))), simplify = FALSE)
# }
# 
# #' Emit a table as RTF via gentlg
# #' @inheritParams formatters::matrix_form
# tt_to_tbldf <- function(tt) {
#   mpf <- matrix_form(tt, indent_rownames = FALSE, expand_newlines = FALSE)
#   strmat <- mf_strings(mpf)
#   nhl <- mf_nlheader(mpf)
#   strbody <- strmat[-(1:nhl), , drop = FALSE]
#   row_type <- apply(strbody, 1, 
#                     function(x) {
#                       nonempty <- nzchar(x)
#                       if(all(!nonempty)) { ## shouldn't ever happen
#                         "EMPTY"
#                       } else if (nonempty[1] && all(!nonempty[-1])) {# only 1st cell nonempty
#                         "HEADER"
#                       } else {
#                         if(x[1] == "N")
#                           "N"
#                         else
#                           "VALUE"
#                       }
#                     })
#   rowdf <- mf_rinfo(mpf)
#   rinds <- mf_lgrouping(mpf)[-(1:nhl)] - mf_nrheader(mpf)
#   indentme <- rowdf$indent[rinds]
#   anbr <- cumsum(!is.na(c(0, head(rowdf$trailing_sep, -1))))[rinds] + 1 ## so it starts at 1
#   roworder <- 1:NROW(rowdf) - (anbr - 1)
#   newrows <- c(0, ifelse(tail(anbr, -1) == head(anbr, -1), 0, 1))
#   
#   tbldf <- as.data.frame(strbody)
#   names(tbldf) <- c("label", paste("col", seq(1, ncol(tbldf) - 1)))
#   cbind(tbldf, data.frame(row_type = row_type,
#                           anbr = anbr,
#                           indentme = indentme,
#                           roworder = roworder, 
#                           newrows = newrows,
#                           stringsAsFactors = FALSE))
# }
# 
# #' @rdname tt_to_tbldf
# #' @param file character(1). File path for output.
# tt_to_tlgrtf <- function(tt, file = "./table") {
#   df <- tt_to_tbldf(tt)
#   mpf <- matrix_form(tt, indent_rownames = FALSE, expand_newlines = FALSE)
#   
#   strmat <- mf_strings(mpf)
#   nhl <- mf_nlheader(mpf)
#   nspancols <- nhl - 1L
#   if(nspancols > 0)
#     csph <- lapply(seq_len(nspancols), function(ii) strmat[ii,])
#   else
#     csph <- NULL
#   gentlg(df,
#          colspan = csph, 
#          file = file,
#          colheader = strmat[nhl,],
#          title = main_title(tt),
#          footers = c(main_footer(tt), prov_footer(tt)))
# }
```

# Load data

NOTE: these datasets were created from source in 'luigi_covid19_data_preprocessing_02202024.Rmd'

```{r}
load("../Data/dt1.RData")
load("../Data/dt_rna.RData")
load("../Data/rnaseq_meta.RData")
load("../Data/rfa.RData")
```

# Clinical Data

## Table 1

```{r}
t1 <- table1( ~.| `Study Group` + AnyDM,
              data = dt1[, -c("Real ID",
                              "Reason_for_Admission",
                              "Adm_DX2",
                              "Admit Date",
                              "Discharge Date")])

t1
```

```{r}
write.table (t1 , 
             file = "../tmp/table1.csv",
             col.names = TRUE, 
             row.names = FALSE,
             # append = TRUE,
             sep = ',')
```

## DX1 Analysis

```{r}
length(unique(rfa$Level1))
level1_n <- rfa[, .(N_Level1 = .N),
                by = list(`Study Group`,
                          Level1)]
level1_n <- dcast.data.table(data = level1_n,
                             Level1 ~ `Study Group`)
level1_n
```

### In-hospital death DX1

```{r}
tmp1 <- rfa[Expired == 1,
           c("Real ID",
             "Study Group",
             "Reason_for_Admission",
             "Age",
             "Race",
             "Level1",
             "Level2",
             "Level3",
             "Level4",
             "Level5",
             "Level6",
             "Level7")]
```

### 90-Day after discharge death DX1

```{r}
tmp2 <- rfa[rfa$`90 Day Mortality` == 1,
           c("Real ID",
             "Study Group",
             "Reason_for_Admission",
             "Age",
             "Race",
             "Level1",
             "Level2",
             "Level3",
             "Level4",
             "Level5",
             "Level6",
             "Level7")]
```

### 90-Day after discharge readmission DX1

```{r}
tmp3 <- rfa[rfa$`90 Day Readmission`== 1,
           c("Real ID",
             "Study Group",
             "Reason_for_Admission",
             "Age",
             "Race",
             "Level1",
             "Level2",
             "Level3",
             "Level4",
             "Level5",
             "Level6",
             "Level7")]
```

```{r}
length(unique(rfa$Level2))
level2_n <- rfa[, .(N_Level2 = .N),
                by = list(`Study Group`,
                          Level1,
                          Level2)]

level2_n <- dcast.data.table(data = level2_n,
                             Level1 + Level2 ~ `Study Group`)
level2_n
```

## Death

### Death vs. COVID

```{r}
addmargins(table(Death = dt1$Expired,
                 COVID = dt1$`Study Group`))
```

```{r}
m1 <- glm(Expired ~ `Study Group`,
          family = binomial,
          data = dt1)
s1 <- summary(m1)
ci1 <- confint(m1)

out1 <- data.table(Response = "Dead",
                   Comparison = "COVID vs. No Covid",
                   OR = round(exp(s1$coefficients[-1, 1]),
                              2),
                   `95% C.I.L.B.` = exp(ci1[2, 1]),
                   `95% C.I.U.B.` = exp(ci1[2, 2]),
                   `p-Value` = ifelse(s1$coefficients[-1, 4] >= 0.001,
                                      round(s1$coefficients[-1, 4],
                                            3),
                                      "<0.001"),
                   sign = ifelse(s1$coefficients[-1, 4] <= 0.05,
                                 ifelse(s1$coefficients[-1, 4] <= 0.01,
                                        "**",
                                        "*"), ""))

out1
```

### Death vs DM, COVID only

```{r}
covid <- droplevels(dt1[`Study Group` == "COVID", ])

chisq.test(covid$Expired,
           covid$AnyDM)

addmargins(table(Death = covid$Expired,
                 AnyDM = covid$AnyDM))

nocovid <- droplevels(dt1[`Study Group` != "COVID", ])
```

### Death vs Obese, COVID only

```{r}
chisq.test(covid$Expired,
           covid$Obese)

addmargins(table(Death = covid$Expired,
                 Obese = covid$Obese))
```

### Death vs BMI, COVID only

```{r}
t.test(covid$BMI ~ covid$Expired)

ggplot(covid,
       aes(x = Expired,
           y = BMI)) +
  geom_boxplot(outlier.shape = "") +
  geom_point(aes(group = `Real ID`),
             shape = 21,
             size = 2,
             position = position_dodge(0.3)) +
  ggtitle("COVID Patients Only") +
  theme_bw()
```

### Death vs Critical Care

Significant

```{r}
addmargins(table(Death = covid$Expired,
                 CrCare = covid$`Critical Care`))

chisq.test(covid$Expired,
           covid$`Critical Care`)

m3 <- glm(Expired ~ `Critical Care`,
          family = binomial,
          data = covid)
s3 <- summary(m3)

out3 <- data.table(Response = "Dead",
                   Comparison = c("ICU: Yes vs. No"),
                   OR = round(exp(s3$coefficients[-1, 1]),
                              2),
                   `95% C.I.L.B.` = exp(s3$coefficients[-1, 1] - 1.96*s3$coefficients[-1, 2]),
                   `95% C.I.U.B.` = exp(s3$coefficients[-1, 1] + 1.96*s3$coefficients[-1, 2]),
                   `p-Value` = ifelse(s3$coefficients[-1, 4] >= 0.001,
                                      round(s3$coefficients[-1, 4], 3),
                                      "<0.001"),
                   sign = ifelse(s3$coefficients[-1, 4] <= 0.05,
                                 ifelse(s3$coefficients[-1, 4] <= 0.01,
                                        "**",
                                        "*"), ""))

out3

m3 <- glm(Expired ~ `Critical Care`,
          family = binomial,
          data = nocovid)
s3 <- summary(m3)

out31 <- data.table(Response = "Dead",
                    Comparison = c("ICU: Yes vs. No"),
                    OR = round(exp(s3$coefficients[-1, 1]),
                               2),
                    `95% C.I.L.B.` = exp(s3$coefficients[-1, 1] - 1.96*s3$coefficients[-1, 2]),
                    `95% C.I.U.B.` = exp(s3$coefficients[-1, 1] + 1.96*s3$coefficients[-1, 2]),
                    `p-Value` = ifelse(s3$coefficients[-1, 4] >= 0.001,
                                       round(s3$coefficients[-1, 4], 3),
                                       "<0.001"),
                    sign = ifelse(s3$coefficients[-1, 4] <= 0.05,
                                  ifelse(s3$coefficients[-1, 4] <= 0.01,
                                         "**",
                                         "*"), ""))

out31
```

### Death vs. COVID \| Diabetes

No significant difference.\
Model with interaction is worse.

```{r}
m2 <- glm(Expired ~ `Study Group` +
            AnyDM,
          family = binomial,
          data = dt1)
s2 <- summary(m2)

out2 <- data.table(Response = "Dead",
                   Comparison = c("COVID vs. No Covid",
                                  "Any DM vs. No DM"),
                   OR = round(exp(s2$coefficients[-1, 1]),
                              2),
                   `95% C.I.L.B.` = exp(s2$coefficients[-1, 1] - 1.96*s2$coefficients[-1, 2]),
                   `95% C.I.U.B.` = exp(s2$coefficients[-1, 1] + 1.96*s2$coefficients[-1, 2]),
                   `p-Value` = ifelse(s2$coefficients[-1, 4] >= 0.001,
                                      round(s2$coefficients[-1, 4], 3),
                                      "<0.001"),
                   sign = ifelse(s2$coefficients[-1, 4] <= 0.05,
                                 ifelse(s2$coefficients[-1, 4] <= 0.01,
                                        "**",
                                        "*"), ""))

out2
```

### Death vs COVID, adjusted for critical care

Both significant in additive model.

```{r}
m4 <- glm(Expired ~ `Study Group` + `Critical Care`,
          family = binomial,
          data = dt1)
s4 <- summary(m4)
ci4 <- confint(m4)

out4 <- data.table(Response = "Dead",
                   Comparison = c("COVID vs. No Covid",
                                  "ICU"),
                   OR = round(exp(s4$coefficients[-1, 1]),
                              2),
                   `95% C.I.L.B.` = exp(ci4[-1, 1]),
                   `95% C.I.U.B.` = exp(ci4[-1, 2]),
                   `p-Value` = ifelse(s4$coefficients[-1, 4] >= 0.001,
                                      round(s4$coefficients[-1, 4],
                                            3),
                                      "<0.001"),
                   sign = ifelse(s4$coefficients[-1, 4] <= 0.05,
                                 ifelse(s4$coefficients[-1, 4] <= 0.01,
                                        "**",
                                        "*"), ""))

out4
```

No significance in interaction model.

```{r}
m41 <- glm(Expired ~ `Study Group`*`Critical Care`,
           family = binomial,
           data = dt1)
summary(m41)
```

### Death vs COVID, adjusted for DM

No significance

```{r}
m5 <- glm(Expired ~ `Study Group` +
            AnyDM,
          family = binomial,
          data = dt1)
summary(m5)
```

### Death vs COVID, adjusted

```{r}
m6 <- glm(Expired ~ `Study Group` +
            AnyDM +
            Age_Category +
            Sex +
            Race +
            BMI,
          family = binomial,
          data = dt1)
summary(m6)
```

### New variable: DM+Obesity

```{r}
chisq.test(dt1$AnyDM,
           dt1$Obese)
```

## ICU

### ICU LOS

```{r}
summary(dt1$`ICU LOS`)
hist(dt1$`ICU LOS`, 200)
hist(log2(dt1$`ICU LOS` + 1), 200)

m7 <- lm(log2(`ICU LOS` + 1) ~ AnyDM +
           Age +
           Sex +
           Race +
           BMI +
           `Critical Care`,
         family = binomial,
         data = dt1)
summary(m7)
```

### Critical care

```{r}
m8 <- glm(`Critical Care` ~ AnyDM +
            Age +
            Sex +
            Race +
            BMI,
          family = binomial,
          data = droplevels(dt1))
summary(m8)
```

## WHO-OSCI

### Desease severity at baseline by WHO-OSCI

if \<5 = mild\
if \>= 5 = severe\
5 = very sick\
6 = intubated\
7 = ventilated\
if 8: dead

```{r}
covid$who_osci_d1_cat <- "Dead"
covid$who_osci_d1_cat[as.numeric(as.character(covid$`WHO-OSCI D1`)) < 8] <- "Severe"
covid$who_osci_d1_cat[as.numeric(as.character(covid$`WHO-OSCI D1`)) < 5] <- "Mild"

covid$who_osci_d1_cat <- factor(covid$who_osci_d1_cat,
                                levels = c("Mild",
                                           "Severe"))
summary(covid$who_osci_d1_cat)
```

```{r}
addmargins(table("WHO-OSCI D1" = dt1$`WHO-OSCI D1`,
                 CC = dt1$`Critical Care`))
```

### WHO-OSCI over time

```{r}
tmp <- dt1[, c("Real ID",
               "Study Group",
               "AnyDM",
               "Expired",
               "WHO-OSCI D1",
               "WHO-OSCI D3",
               "WHO-OSCI D7",
               "WHO-OSCI D14",
               "WHO-OSCI D21",
               "WHO-OSCI D28",
               "WHO-OSCI DISCHARGE")]

tmp <- melt.data.table(tmp,
                       id.vars = c(1:4),
                       variable.name = "Day",
                       value.name = "WHO_OSCI")
tmp$Day <- factor(tmp$Day,
                  levels = unique(tmp$Day))

tmp <- tmp[!is.na(tmp$WHO_OSCI), ]
ggplot(tmp,
       aes(x = Day,
           y = WHO_OSCI,
           group = `Real ID`,
           color = `Real ID`)) +
  facet_wrap(~ AnyDM) +
  geom_line(position = position_dodge(0.3)) +
  theme_bw() +
  theme(legend.position = "none") +
  theme(axis.text.x = element_text(angle = 45,
                                   hjust = 1))
```

```{r}
table(dt1$`WHO-OSCI D1`,
      dt1$Expired)
```

### Table 2

```{r}
t2 <- dcast.data.table(`Real ID` ~ Day,
                       value.var = "WHO_OSCI",
                       data = tmp[Expired == "Yes", ])

write.csv(t2,
          file = "tmp/table2.csv",
          row.names = FALSE)

t2
```

```{r}
dt1[,
    .(N = .N,
      mu_los = mean(LOS),
      sem_los = sd(LOS)/sqrt(.N)),
    by = c("Study Group")]

dt1[,
    .(N = .N,
      mu_los = mean(LOS),
      sem_los = sd(LOS)/sqrt(.N)),
    by = c("Expired")]

dt1[,
    .(N = .N,
      mu_los = mean(LOS),
      sem_los = sd(LOS)/sqrt(.N)),
    by = c("Expired",
           "Study Group")]

dt1[LOS > 1,
    .(N = .N,
      mu_los = mean(LOS),
      sem_los = sd(LOS)/sqrt(.N)),
    by = c("Expired",
           "Study Group")]
```

```{r}
dt1$LOS[dt1$`Real ID` == 771982]
```

# ELISA

```{r}
dt_elisa <- data.table(dt1[, c("Real ID",
                               "Study Group",
                               "AnyDM")],
                       dt1[, `G-CSF (CSF-3)`:`ACE2 (ng/ml)`])

n_col <- ncol(dt1[, `G-CSF (CSF-3)`:`ACE2 (ng/ml)`])

## Remove patients with no ELISA data
# ndx <- which(rowSums(is.na(dt1[, `G-CSF (CSF-3)`:`ACE2 (ng/ml)`])) == n_col)
# ndx
# dt_elisa <- dt_elisa[-ndx, ]

# Remove patients with most ELISA data missing
dt_elisa <- dt_elisa[!is.na(dt_elisa$`G-CSF (CSF-3)`), ]

addmargins(table(COVID = dt_elisa$`Study Group`,
                 AnyDM = dt_elisa$AnyDM))

summary(dt_elisa)
```

## ELISA long format

```{r}
dt_elisa_l <- melt.data.table(data = dt_elisa,
                              id.vars = 1:3,
                              measure.vars = 4:ncol(dt_elisa),
                              variable.name = "Protein",
                              value.name = "Concentration[pg/mL]")

dt_elisa_l[, grp := factor(paste(`Study Group`,
                                 AnyDM,
                                 sep = "_"),
                           levels = c("No COVID_0",
                                      "No COVID_1",
                                      "COVID_0",
                                      "COVID_1"),
                           labels = c("No COVID/No DM",
                                      "No COVID/DM",
                                      "COVID/No DM",
                                      "COVID/DM"))]

dt_elisa_l$Log2Conc <- log2(dt_elisa_l$`Concentration[pg/mL]` + 1)

# Remove NAs
dt_elisa_l <- dt_elisa_l[!is.na(Log2Conc), ]
```

### Histograms

```{r,fig.height=10,fig.width=10}
p1 <- ggplot(dt_elisa_l,
             aes(x = `Concentration[pg/mL]`)) +
  facet_wrap(~ Protein,
             scale = "free") +
  geom_histogram(bins = 30) +
  theme_bw()

tiff(filename = "../tmp/elisa_hist.tiff",
     height = 10,
     width = 10,
     units = 'in',
     res = 600,
     compression = "lzw+p",)
print(p1)
graphics.off()

p1
```

### Boxplot

```{r,fig.height=10,fig.width=10}
p1 <- ggplot(dt_elisa_l,
             aes(x = grp,
                 y = `Concentration[pg/mL]`)) +
  facet_wrap(~ Protein,
             scale = "free_y") +
  geom_boxplot(outlier.shape = "") +
  geom_point(aes(fill = grp,
                 group = `Real ID`,),
             shape = 21,
             size = 3,
             alpha = 0.7,
             position = position_dodge(0.3)) +
  theme_bw() +
  theme(legend.position = "none",
        axis.text.x = element_text(angle = 45,
                                   hjust = 1))

tiff(filename = "../tmp/elisa_boxplot.tiff",
     height = 10,
     width = 10,
     units = 'in',
     res = 600,
     compression = "lzw+p",)
print(p1)
graphics.off()

p1
```

## Log2 protein concentrations

### Histogram

```{r,fig.height=10,fig.width=10}
ggplot(dt_elisa_l,
       aes(x = Log2Conc)) +
  facet_wrap(~ Protein,
             scale = "free") +
  geom_histogram(bins = 30) +
  theme_bw()

tiff(filename = "../tmp/elisa_log2_hist.tiff",
     height = 10,
     width = 10,
     units = 'in',
     res = 600,
     compression = "lzw+p",)
print(p1)
graphics.off()

p1
```

### Boxplots

```{r,fig.height=10,fig.width=10}
p1 <- ggplot(dt_elisa_l,
             aes(x = grp,
                 y = Log2Conc)) +
  facet_wrap(~ Protein,
             scale = "free_y") +
  geom_boxplot(outlier.shape = "") +
  geom_point(aes(fill = grp,
                 group = `Real ID`,),
             shape = 21,
             size = 2,
             alpha = 0.7,
             position = position_dodge(0.3)) +
  scale_x_discrete("") +
  scale_y_continuous("Concentration[pg/mL]",
                     breaks = 0:15,
                     labels = 2^(0:15) - 1) +
  theme_bw() +
  theme(legend.position = "none",
        axis.text.x = element_text(angle = 60,
                                   hjust = 1))

tiff(filename = "../tmp/elisa_log2_boxplot.tiff",
     height = 10,
     width = 10,
     units = 'in',
     res = 600,
     compression = "lzw+p",)
print(p1)
graphics.off()

p1
```

```{r}
# out <- list()
# for (i in 4:ncol(dt_elisa)) {
#   
#   tmp <- data.table(id = dt_elisa$`Real ID`,
#                     covid = dt_elisa$`Study Group`,
#                     dm = factor(dt_elisa$AnyDM,
#                                 levels = 0:1,
#                                 labels = c("No DM",
#                                            "DM")),
#                     resp = log10(dt_elisa[, c(i), with = FALSE][[1]] + 0.01))
# 
#   p1 <- ggplot(tmp,
#                aes(x = covid,
#                    y = resp)) +
#     facet_wrap(~ dm) +
#     geom_boxplot(outlier.shape = "") +
#     geom_point(aes(fill = covid,
#                    group = id,),
#                shape = 21,
#                size = 3,
#                alpha = 0.7,
#                position = position_dodge(0.3)) +
#     ggtitle(colnames(dt_elisa)[i]) +
#     theme_bw()
#   print(p1)
#   
#   s1 <- summary(lm(resp ~ covid*dm,
#                    data = tmp))
#   
#   out[[i- 3]] <- data.table(protein = colnames(dt_elisa)[i],
#                             comparison = c("COVID - non-COVID",
#                                            "DM - non-DM",
#                                            "COVID*DM Interaction"),
#                             diff = round(s1$coefficients[-1, 1], 3),
#                             se_diff = round(s1$coefficients[-1, 2], 3),
#                             pval = round(s1$coefficients[-1, 4], 3))
# }
```

## ELISA Two Part model

```{r}
foo <- function(dx) {
  return(tpm(formula_part1 = Log2Conc ~ `Study Group`*AnyDM,
             formula_part2 = NULL,
             link_part1 = "logit",
             family_part2 = gaussian(link = "identity"),
             data = dx))
}

m1 <- dt_elisa_l[, .(models = list(try(tpm(formula_part1 = Log2Conc ~ grp,
                                           formula_part2 = NULL,
                                           link_part1 = "logit",
                                           family_part2 = gaussian(link = "identity"),
                                           data = data.table(Log2Conc = Log2Conc,
                                                         grp = grp))))),
                 by = Protein]
```

```{r}
# m1$Protein
#
# mm1 <- m1$models[[1]]
# s1 <- summary(mm1)
# s1
# 
# est_ame <- AME(object = mm1,
#                se.method = "delta")
# 
# est_ame <- AME(object = mm1,
#                se.method = "bootstrap")
# est_ame
```

```{r}
out <- lapply(m1$models,
              function(a) {
                try({
                  s1 <- summary(a)
                  
                  # Average Marginal Effect
                  # est_ame <- AME(object = a,
                  #                se.method = "bootstrap")
                  # NOTE: convergence issues iwth bootstrap!
                  
                  est_ame <- AME(object = a,
                                 se.method = "delta")
                  
                  rbindlist(list(data.table(Model = "Logistic",
                                            Comparisons = c("COVID/No DM - No COVID/No DM",
                                                            "No COVID/DM - No COVID/No DM",
                                                            "COVID/DM - No COVID/No DM"),
                                            Estimated = "Odds Ratio of Finding the Protein (non-zero)",
                                            Estimate = exp(s1$Firstpart.model$coefficients[-1, 1]),
                                            CILB = exp(s1$Firstpart.model$coefficients[-1, 1] - 
                                                         1.96*s1$Firstpart.model$coefficients[-1, 2]),
                                            CIUB = exp(s1$Firstpart.model$coefficients[-1, 1] + 
                                                         1.96*s1$Firstpart.model$coefficients[-1, 2]),
                                            pVal = s1$Firstpart.model$coefficients[-1, 4]),
                                 
                                 data.table(Model = "Gaussian",
                                            Comparisons = c("COVID/No DM - No COVID/No DM",
                                                            "No COVID/DM - No COVID/No DM",
                                                            "COVID/DM - No COVID/No DM"),
                                            Estimated = "Log10(Concentration [pg/mL] + 1) Difference",
                                            Estimate = s1$Secondpart.model$coefficients[-1, 1],
                                            CILB = s1$Secondpart.model$coefficients[-1, 1] - 
                                              1.96*s1$Secondpart.model$coefficients[-1, 2],
                                            CIUB = s1$Secondpart.model$coefficients[-1, 1] + 
                                              1.96*s1$Secondpart.model$coefficients[-1, 2],
                                            pVal = s1$Secondpart.model$coefficients[-1, 4]),
                                 data.table(Model = "Average Marginal Effect",
                                            Comparisons = c("COVID/No DM - No COVID/No DM",
                                                            "No COVID/DM - No COVID/No DM",
                                                            "COVID/DM - No COVID/No DM"),
                                            Estimated = "Log10(Concentration [pg/mL] + 1) Difference",
                                            Estimate = est_ame$dydx,
                                            CILB = est_ame$CI.L,
                                            CIUB = est_ame$CI.U,
                                            pVal = est_ame$pvalue)))
                })
              })

out <- do.call("rbind",
               out)

out <- data.table(Protein = rep(m1$Protein,
                                each = 9),
                  out)

write.csv(out,
          "../tmp/two_part_models_elisa.csv",
          row.names = FALSE)
head(out)
```

```{r}
out$Estimate[is.infinite(out$CIUB)] <-  NA
out$CILB[is.infinite(out$CIUB)] <-  NA
out$CIUB[is.infinite(out$CIUB)] <- NA
```

### Logistic

```{r,fig.height=9,fig.width=7}
p1 <- ggplot(data = out[Model == "Logistic", ],
             aes(x = Estimate,
                 xmin = CILB,
                 xmax = CIUB,
                 y = Comparisons,
                 fill = Comparisons)) +
  facet_wrap(~ Protein,
             scale = "free_x",
             ncol = 3) +
  geom_errorbarh(height = 0.4) +
  geom_vline(xintercept = 1,
             linetype = "dashed") +
  geom_point(shape = 21,
             size = 3) +
  scale_x_continuous("Odds Ratio of Observing Non-Zero") +
  ggtitle("Logistic Part of the Model") +
  theme_bw() +
  theme(legend.position = "none")

tiff(filename = "../tmp/elisa_two_part_logist.tiff",
     height = 9,
     width = 7,
     units = 'in',
     res = 600,
     compression = "lzw+p")
print(p1)
graphics.off()

p1
```

### Gaussian

```{r,fig.height=9,fig.width=7}
p1 <- ggplot(data = out[Model == "Gaussian", ],
             aes(x = Estimate,
                 xmin = CILB,
                 xmax = CIUB,
                 y = Comparisons,
                 fill = Comparisons)) +
  facet_wrap(~ Protein,
             scale = "free_x",
             ncol = 3) +
  geom_errorbarh(height = 0.3) +
  geom_vline(xintercept = 0,
             linetype = "dashed") +
  geom_point(shape = 21,
             size = 3) +
  scale_x_continuous("Log10(Concentration [pg/mL] + 1) Difference") +
  ggtitle("Gaussian Part of the Model") +
  theme_bw() +
  theme(legend.position = "none")

tiff(filename = "../tmp/elisa_two_part_gaussian.tiff",
     height = 9,
     width = 7,
     units = 'in',
     res = 600,
     compression = "lzw+p")
print(p1)
graphics.off()

p1
```

### Average Marginal Effect

```{r,fig.height=9,fig.width=7}
p1 <- ggplot(data = out[Model == "Average Marginal Effect", ],
             aes(x = Estimate,
                 xmin = CILB,
                 xmax = CIUB,
                 y = Comparisons,
                 fill = Comparisons)) +
  facet_wrap(~ Protein,
             scale = "free_x",
             ncol = 3) +
  geom_errorbarh(height = 0.3) +
  geom_vline(xintercept = 0,
             linetype = "dashed") +
  geom_point(shape = 21,
             size = 3) +
  scale_x_continuous("Log10(Concentration [pg/mL] + 1) Difference") +
  ggtitle("Average Marginal Effect") +
  theme_bw() +
  theme(legend.position = "none")

tiff(filename = "../tmp/elisa_two_part_average_marginal_effect.tiff",
     height = 9,
     width = 7,
     units = 'in',
     res = 600,
     compression = "lzw+p")
print(p1)
graphics.off()

p1
```

## Factor analysis

Source: <https://personality-project.org/r/psych/HowTo/factor.pdf>

### Data

```{r}
dt_fa <- as.data.frame(log2(dt_elisa[, `G-CSF (CSF-3)`:`ACE2 (ng/ml)`] + 1)) 
rownames(dt_fa) <- dt_elisa$`Real ID`

describe(dt_fa)
```

### Plot pairs

```{r,fig.height=20,fig.width=20}
tiff(filename = "../tmp/elisa_pairs.tiff",
     height = 20,
     width = 20,
     units = 'in',
     res = 600,
     compression = "lzw+p")
pairs.panels(dt_fa)
graphics.off()

pairs.panels(dt_fa)
```

### Correlation coefficients

```{r}
psych::lowerCor(dt_fa)
```

### Correlation Heatmap

```{r,fig.height=10,fig.width=10}
tiff(filename = "../tmp/elisa_cor_heatmap.tiff",
     height = 10,
     width = 10,
     units = 'in',
     res = 600,
     compression = "lzw+p")
psych::corPlot(dt_fa)
graphics.off()

psych::corPlot(dt_fa)
```

### Diagram

```{r}
fa.diagram(dt_fa)
```

### Test number of factors

```{r}
m1 <- fa.parallel(x = dt_fa)

plot(m1)

m1
```

### Very Simple Structure

```{r}
vss(dt_fa)
```

### ICLUST algorithm

Revelle, 1979

```{r}
psych::iclust(r.mat = dt_fa,
              nclusters = 1)
```

### OMEGA

```{r}
omega(dt_fa)
```

### bassAckward

```{r,fig.height=10,fig.width=7}
bassAckward(dt_fa,
            nfactors = 3)
```

### Alpha

```{r}
alpha(dt_fa)

## This code doesn't work
# my_keys <- make.keys(nvars = 23,
#                     keys.list = list(first = c(1, 3, 5),
#                                      second = c(2, 3, 10)))
# my_keys
# 
# my_scores <- scoreItems(keys = my_keys,
#                         items = dt_fa)
# my_scores
```

### FA

#### Model

```{r}
m1 <- psych::fa(r = dt_fa,
                fm = "fa",
                nfactors = 3)
m1
summary(m1)

target.rot(m1)
```

#### Plot factors

```{r,fig.height=5,fig.width=5}
plot(m1)
```

#### Biplot

```{r,fig.height=10,fig.width=10}
psych::biplot.psych(m1)
```

#### Plot diagram

```{r,fig.height=10,fig.width=6}
fa.diagram(m1)
```

```{r}
m1$scores

m1$loadings
```

# CONTINUE HERE 3/21/24

## ELISA binarized

```{r}
# dt_elisa_bin <-  copy(dt_elisa)
# dt_elisa_bin[, 4:ncol(dt_elisa)] <- lapply(dt_elisa[, 4:ncol(dt_elisa)],
#                                            function (a) {
#                                              return(as.numeric(a > 0))
#                                            })
# dt_elisa_bin
```

## LM results

```{r}
res <- rbindlist(out)
# res$pval[res$pval == 0] <- "<0.001"
res
```

## ELISA PCA

### PCA

```{r}
m1 <- prcomp(dt_elisa[, -c(1:3)],
             center = TRUE,
             scale. = TRUE)
summary(m1)
```

### Scores (i.e. points)

```{r}
dt_scr <- data.table(m1$x[, 1:2])
dt_scr$covid <- dt_elisa$`Study Group`
dt_scr$dm <- factor(dt_elisa$AnyDM,
                    levels = 0:1,
                    labels = c("No DM",
                               "DM"))
head(dt_scr)
```

### Loading (i.e. arrows)

```{r}
dt_rot <- as.data.frame(m1$rotation[, 1:2])
dt_rot$feat <- rownames(dt_rot)
dt_rot <- data.table(dt_rot)
dt_rot
```

### Axis labels

```{r}
u.axis.labs <- paste(colnames(dt_rot)[1:2], 
                     sprintf('(%0.1f%% explained var.)', 
                             100*m1$sdev[1:2]^2/sum(m1$sdev^2)))
u.axis.labs
```

### Biplot by COVID

```{r,fig.width=8,fig.height=8}
grp <- dt_scr$covid
scl <- 30

p1 <- ggplot(data = dt_rot,
             aes(x = PC1,
                 y = PC2)) +
  coord_equal() +
  geom_point(data = dt_scr,
             aes(fill = grp),
             shape = 21,
             size = 3,
             alpha = 0.7) +
  geom_segment(aes(x = 0,
                   y = 0,
                   xend = scl*PC1,
                   yend = scl*PC2),
               arrow = arrow(length = unit(1/2, 'picas')),
               color = "black",
               size = 0.5) +
  geom_text(aes(x = 1.1*scl*PC1,
                y = 1.1*scl*PC2,
                label = dt_rot$feat),
            size = 2,
            hjust = 0.5) +
  scale_x_continuous(u.axis.labs[1]) +
  scale_y_continuous(u.axis.labs[2]) +
  scale_fill_discrete(name = "Group") +
  ggtitle("Biplot") +
  theme_bw() +
  theme(plot.title = element_text(hjust = 0.5,
                                  size = 20))
p1
# tiff(filename = "tmp/pca_biplot.tiff",
#      height = 10,
#      width = 10,
#      units = 'in',
#      res = 300,
#      compression = "lzw+p")
# print(p1)
# graphics.off()
```

### Biplot by COVID

```{r,fig.width=8,fig.height=8}
grp <- dt_scr$dm
scl <- 30

p1 <- ggplot(data = dt_rot,
             aes(x = PC1,
                 y = PC2)) +
  coord_equal() +
  geom_point(data = dt_scr,
             aes(fill = grp),
             shape = 21,
             size = 3,
             alpha = 0.7) +
  geom_segment(aes(x = 0,
                   y = 0,
                   xend = scl*PC1,
                   yend = scl*PC2),
               arrow = arrow(length = unit(1/2, 'picas')),
               color = "black",
               size = 0.5) +
  geom_text(aes(x = 1.1*scl*PC1,
                y = 1.1*scl*PC2,
                label = dt_rot$feat),
            size = 2,
            hjust = 0.5) +
  scale_x_continuous(u.axis.labs[1]) +
  scale_y_continuous(u.axis.labs[2]) +
  scale_fill_discrete(name = "Group") +
  ggtitle("Biplot") +
  theme_bw() +
  theme(plot.title = element_text(hjust = 0.5,
                                  size = 20))
p1
# tiff(filename = "tmp/pca_biplot.tiff",
#      height = 10,
#      width = 10,
#      units = 'in',
#      res = 300,
#      compression = "lzw+p")
# print(p1)
# graphics.off()
```

# RNA-seq

<https://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html>

## Merge RNA-seq meta-data with clinical

```{r}
rnaseq_meta_clin <- merge(rnaseq_meta,
                          dt1[, c("Real ID",
                                  "Study Group",
                                  "Expired",
                                  "Critical Care",
                                  "Age",
                                  "Sex",
                                  "Race",
                                  "BMI",
                                  "AnyDM")],
                          by = "Real ID")

names(rnaseq_meta_clin)[3] <- "Cohort"
rnaseq_meta_clin$Cohort <- gsub(x = rnaseq_meta_clin$Cohort,
                                pattern = " ",
                                replacement = "_")
rnaseq_meta_clin$Cohort <- factor(rnaseq_meta_clin$Cohort,
                                  levels = c("No_COVID",
                                             "COVID"))
```

## Fragments Per Million (FPM)

### Sequencing depth

```{r}
seq_d <- colSums(dt_rna[, -c(1:11)])
hist(seq_d, 20)
```

### DESeq: all genes

```{r}
ndx <- which(colnames(dt_rna) %in% rnaseq_meta_clin$sample_names)

tmp <- as.matrix(dt_rna[, ndx, with = FALSE])
rownames(tmp) <- dt_rna$gene_id

dds_all <- DESeqDataSetFromMatrix(countData = tmp, 
                                  colData = rnaseq_meta_clin,
                                  design = ~ 1)
dds_all
```

### FPM

```{r}
dt_fpm <- fpm(object = dds_all,
              robust = FALSE)
dt_fpm[1:10, 1:5]
colSums(dt_fpm)[1:5]
```

```{r}
tmp5

xx <- data.table(fpm = dt_fpm[which(rownames(dt_fpm) == "ENSG00000165949"), ],
                 sample_names = colnames(dt_fpm))

xx <- merge(xx,
            rnaseq_meta_clin,
            by = "sample_names")
xx

boxplot(log10(xx$fpm + 1) ~ xx$Cohort)
```

## Protein coding genes only

```{r}
dt_rna_cod <- dt_rna[gene_biotype == "protein_coding",]
```

## Select samples with clinical information

```{r}
ndx <- which(colnames(dt_rna_cod) %in% rnaseq_meta_clin$sample_names)
```

## DESeq

```{r}
tmp <- as.matrix(dt_rna_cod[, ndx, with = FALSE])
rownames(tmp) <- dt_rna_cod$gene_id

dds_cod <- DESeqDataSetFromMatrix(countData = tmp, 
                                  colData = rnaseq_meta_clin,
                                  ~ 1)
```

## Filter out genes with small number of hits

Keep the genes with at least 5 out of 92 samples having at least 10 hits each

```{r}
# Number of samples with at least 10 hits in at least X samples
min_n_hits <- c(1, 5, 10, 20, 50)

l2 <- list()

for (j in 1:length(min_n_hits)) {
  
  l1 <- list()
  
  for (i in 1:length(ndx)) {
    
    l1[[i]] <- data.table(`Min N of hits per sample` = min_n_hits[j],
                          `N samples with min hits or more` = i,
                          `N genes` = sum(rowSums(counts(dds_cod) >= j) >= i))
  }
  l2[[j]] <- do.call("rbind",
                     l1)
}

ll <- do.call("rbind",
              l2)
ll$`Min N of hits per sample` <- factor(ll$`Min N of hits per sample`,
                                        levels = min_n_hits)
```

```{r}
ll$lw <- 0.5
ll$lw[ll$`Min N of hits per sample` == 10] <- 1.5

ggplot(ll,
       aes(x = `N samples with min hits or more`,
           y = `N genes`,
           group = `Min N of hits per sample`,
           color = `Min N of hits per sample`)) +
  geom_line(linewidth = ll$lw) +
  geom_vline(xintercept = 10,
             linetype = "dashed",
             color = "red") +
  scale_x_continuous(breaks = seq(0, 92, 5)) +
  theme_bw()
```

```{r}
keep <- rowSums(counts(dds_cod) >= 10) >= 10
sum(keep)

dds_cod <- dds_cod[keep, ]
```

## COVID/non-COOVID

### Restructure design (model)

```{r}
design(dds_cod) <- ~ Cohort
```

### Run DESeq

```{r}
deseq_cod <- DESeq(dds_cod)
resultsNames(deseq_cod)
```

### Results

```{r}
res <- results(deseq_cod)
res[1:10, ]
rownames(res)[1:10]
```

```{r}
out <- data.table(gene_id = rownames(res),
                  baseMean = res$baseMean,
                  log2FoldChange = res$log2FoldChange,
                  log2FoldChangeSE = res$lfcSE,
                  stat = res$stat,
                  pvalue = res$pvalue,
                  padj = res$padj)

out <- merge(dt_rna_cod[, c("gene_name",
                            "gene_id")],
             out,
             by = "gene_id")

out <- out[!is.na(padj), ]

setorder(out, 
         padj)


head(out)

write.csv(out,
          file = "tmp/deseq2_cod_covid_nocovid.CSV",
          row.names = FALSE)
```

```{r}
tmp5 <- out[abs(log2FoldChange) > 1 &
              padj < 0.05, ]
# 
# tmp5_up <- out[log2FoldChange > 1 &
#               padj < 0.05, ]
# 
# tmp5_down <- out[log2FoldChange < -1 &
#               padj < 0.05, ]
```

### Calculate the mean expressions of the selected genes

```{r}
tmp <- dt_rna_cod[gene_name %in% tmp5$gene_name, ]
tmp <- melt.data.table(data = tmp,
                       id.vars = 1,
                       measure.vars = 12:ncol(tmp),
                       variable.name = "sample_names",
                       value.name = "Hits")

tmp <- merge(tmp,
             rnaseq_meta_clin,
             by = "sample_names")
```

```{r,fig.width=10,fig.height=10}
p1 <- ggplot(tmp,
             aes(x = Cohort,
                 y = log2(Hits + 1))) +
  facet_wrap(~ gene_name,
             scale = "free_y") +
  geom_boxplot(width = 0.3,
               outlier.shape = "") +
  geom_point(aes(fill = AnyDM,
                 group = sample_names),
             size = 2,
             shape = 21,
             alpha = 0.5,
             position = position_dodge(0.3)) +
  scale_y_continuous("Number of hits",
                     breaks = seq(from = 1,
                                  to = 20,
                                  by = 2),
                     labels = (2^seq(from = 1,
                                     to = 20,
                                     by = 2)) - 1) +
  scale_fill_discrete("Any DM",
                      breaks = c(0, 1),
                      labels = c("No",
                                 "Yes")) +
  ggtitle(paste(c("Genes significantly differentially expressed in COVID-19 vs non-COVID patients.",
                  "Difference of the means of log2[hits+1] > 1 and adjusted p-value < 0.05"),
                collapse = "\n"))+
  theme_bw() +
  theme(legend.position = "bottom")

tiff(filename = "tmp/fig1_sign_genes_cod.tiff",
     height = 10,
     width = 10,
     units = 'in',
     res = 600,
     compression = "lzw+p")
print(p1)
graphics.off()

p1
```

```{r,fig.height=5,fig.width=8}
p2 <- ggplot(tmp[gene_name == "IFI27", ],
             aes(x = Cohort,
                 y = log2(Hits + 1))) +
  geom_boxplot(width = 0.3,
               outlier.shape = "") +
  geom_point(aes(fill = Expired,
                 group = sample_names),
             size = 3,
             shape = 21,
             alpha = 0.8,
             position = position_dodge(0.3)) +
  scale_y_continuous("Number of hits",
                     breaks = seq(from = 1,
                                  to = 20,
                                  by = 2),
                     labels = (2^seq(from = 1,
                                     to = 20,
                                     by = 2)) - 1) +
  scale_fill_discrete("In-Hospital Death",
                      breaks = c(0, 1),
                      labels = c("No",
                                 "Yes")) +
  ggtitle(paste(c("IFI27 expression in COVID-19 vs non-COVID patients.",
                  "Difference of the means of log2[hits+1] > 1",
                  "and adjusted p-value < 0.05"),
                collapse = "\n")) +
  theme_bw() +
  theme(legend.position = "bottom")

p3 <- ggplot(tmp[gene_name == "IFI27", ],
             aes(x = Cohort,
                 y = log2(Hits + 1))) +
  geom_boxplot(width = 0.3,
               outlier.shape = "") +
  geom_point(aes(fill = `Critical Care`,
                 group = sample_names),
             size = 3,
             shape = 21,
             alpha = 0.8,
             position = position_dodge(0.3)) +
  scale_y_continuous("Number of hits",
                     breaks = seq(from = 1,
                                  to = 20,
                                  by = 2),
                     labels = (2^seq(from = 1,
                                     to = 20,
                                     by = 2)) - 1) +
  scale_fill_discrete("Critical Care",
                      breaks = c(0, 1),
                      labels = c("No",
                                 "Yes")) +
  ggtitle(paste(c("IFI27 expression in COVID-19 vs non-COVID patients.",
                  "Difference of the means of log2[hits+1] > 1",
                  "and adjusted p-value < 0.05"),
                collapse = "\n")) +
  theme_bw() +
  theme(legend.position = "bottom")

tiff(filename = "tmp/fig2_ifi27_covid_noncovid.tiff",
     height = 6,
     width = 9,
     units = 'in',
     res = 600,
     compression = "lzw+p")
grid.arrange(p2,
             p3,
             nrow = 1)
# grid.arrange(p2,
#              p3,
#              nrow = 1,
#              top = textGrob(paste(c("IFI27 expression in COVID-19 vs non-COVID patients.",
#                                     "Difference of the means of log2[hits+1] > 1",
#                                     "and adjusted p-value < 0.05"),
#                                   collapse = "\n"),
#                             gp = gpar(fontsize = 20,
#                                       font = 3)))
graphics.off()

grid.arrange(p2,
             p3,
             nrow = 1)
```

### Table 3: COVID/non-COVID significant DEs

```{r}
# mu <- tmp[, .(mu = exp(mean(log(Hits + 1))) - 1,
#               med = median(Hits)),
#               by = list(gene_name,
#                         Cohort)]
# mu

med <- tmp[, .(med = median(Hits)),
           by = list(gene_name,
                     Cohort)]
med <- dcast.data.table(data = med,
                        gene_name ~ Cohort,
                        value.var = "med")
med

t3 <- data.table(gene_name = tmp5$gene_name,
                 `Log2 Fold Change` = round(tmp5$log2FoldChange, 2),
                 `SE of Log2 Fold Change` = round(tmp5$log2FoldChangeSE, 2),
                 `Adjusted p-Value` = round(tmp5$padj, 3))

t3 <- merge(med,
            t3,
            by = "gene_name")

names(t3)[1:3] <- c("Gene",
                    "Median hits in non-COVID",
                    "Median hits in COVID")

write.csv(t3,
          file = "tmp/t3.csv",
          row.names = FALSE)
t3
```

### Reactom Pathway Analysis

#### Top 18 genes

Source: <https://stackoverflow.com/questions/71131883/r-how-do-i-convert-gene-symbols-to-entrez-ids>

##### Map gene names to Entrez IDs

```{r}
dt_rpa1 <- merge(dt_rna[!is.na(entrez_id), 
                        c("gene_name",
                          "entrez_id")],
                 tmp5[, c("gene_name",
                          "log2FoldChange")])
```

##### Run Reactom PA

```{r}
rpa1 <- ReactomePA::enrichPathway(gene = dt_rpa1$entrez_id,
                                  readable = TRUE)

write.csv(rpa1@result,
          file = "tmp/Pathway Analysis of COVID19 18 differentially expresed genes.csv")

rpa1@result
```

##### Barplot of pathways

NOTE: by default only shows pathways with adjusted p-value \<0.05

```{r}
barplot(rpa1,
        showCategory = 4,
        title = "")
```

```{r}
# viewPathway(pathName = rpa1@result$Description[1])

# for(i in 1:length(rpa1@result$Description)) {
#   viewPathway(pathName = rpa1@result$Description[i])
# }

geneList1 <- dt_rpa1$log2FoldChange
names(geneList1) <- dt_rpa1$entrez_id
geneList1

# tiff(filename = paste0("tmp/",
#                        rpa1@result$Description[1],
#                        ".tiff"),
#      height = 8,
#      width = 8,
#      units = 'in',
#      res = 300,
#      compression = "lzw+p")
viewPathway(pathName = rpa1@result$Description[17],
            readable = TRUE,
            foldChange = geneList1)
# graphics.off()

# for(i in 1:length(rpa1@result$Description)) {
# for(i in 1:10) {
#   try({
#     tiff(filename = paste0("tmp/",
#                            rpa1@result$Description[i],
#                            ".tiff"),
#          height = 8,
#          width = 8,
#          units = 'in',
#          res = 600,
#          compression = "lzw+p")
#     viewPathway(pathName = rpa1@result$Description[i],
#                 readable = TRUE,
#                 foldChange = geneList1)
#     graphics.off()
#   })
# }
```

##### Enrichment analysis based on the DisGeNET

<https://www.disgenet.org/>

COVID19 Data on DisGeNET:\
<https://www.disgenet.org/covid/diseases/summary/>

DisGeNET is a discovery platform containing one of the largest publicly available collections of genes and variants associated to human diseases ( Piñero et al., 2021; Piñero et al., 2019; Piñero et al., 2016; Piñero et al., 2015). DisGeNET integrates data from expert curated repositories, GWAS catalogues, animal models and the scientific literature. DisGeNET data are homogeneously annotated with controlled vocabularies and community-driven ontologies. Additionally, several original metrics are provided to assist the prioritization of genotype--phenotype relationships.

The current version of DisGeNET (v7.0) contains 1,134,942 gene-disease associations (GDAs), between 21,671 genes and 30,170 diseases, disorders, traits, and clinical or abnormal human phenotypes, and 369,554 variant-disease associations (VDAs), between 194,515 variants and 14,155 diseases, traits, and phenotypes.

```{r}
tt1 <- DOSE::enrichDGN(dt_rpa1$entrez_id)
tt1@result

# barplot(tt1)
```

#### Top 64 genes

```{r}
tmp6 <- out[abs(log2FoldChange) > 0.5 &
              padj < 0.1, ]
```

```{r}
gene_list_2 <- mapIds(org.Hs.eg.db,
                      keys = tmp6$gene_name,
                      column = "ENTREZID",
                      keytype = "SYMBOL")
gene_list_2
```

```{r}
rpa2 <- ReactomePA::enrichPathway(gene = gene_list_2,
                                  readable = TRUE)

write.csv(rpa2@result,
          file = "tmp/Pathway Analysis of COVID19 64 differentially expresed genes.csv")

rpa2@result
```

## AnyDM/no DM

```{r}
design(dds_cod) <- ~ AnyDM
```

### Run DESeq

```{r}
deseq_cod_dm <- DESeq(dds_cod)

resultsNames(deseq_cod_dm)
```

### DESeq results

```{r}
res_dm <- results(deseq_cod_dm)
res_dm[1:10, ]
rownames(res_dm)[1:10]
```

```{r}
out_dm <- data.table(gene_id = rownames(res_dm),
                     baseMean = res_dm$baseMean,
                     log2FoldChange = res_dm$log2FoldChange,
                     log2FoldChangeSE = res_dm$lfcSE,
                     stat = res_dm$stat,
                     pvalue = res_dm$pvalue,
                     padj = res_dm$padj)

out_dm <- merge(dt_rna_cod[, c("gene_name",
                               "gene_id")],
                out_dm,
                by = "gene_id")

out_dm <- out_dm[!is.na(padj), ]

setorder(out_dm, 
         padj)


head(out_dm)

write.csv(out_dm,
          file = "tmp/deseq2_cod_dm.CSV",
          row.names = FALSE)
```

```{r}
tmp7 <- out_dm[abs(log2FoldChange) > 1 &
                 padj < 0.05, ]
```

### Calculate the mean expressions of the selected genes

```{r}
tmp <- dt_rna_cod[gene_name %in% tmp7$gene_name, ]
tmp <- melt.data.table(data = tmp,
                       id.vars = 1,
                       measure.vars = 12:ncol(tmp),
                       variable.name = "sample_names",
                       value.name = "Hits")

tmp <- merge(tmp,
             rnaseq_meta_clin,
             by = "sample_names")
```

### Table 4: DM/no DM significant DEs

```{r}
med <- tmp[, .(med = median(Hits)),
           by = list(gene_name,
                     Cohort)]
med <- dcast.data.table(data = med,
                        gene_name ~ Cohort,
                        value.var = "med")
med

t4 <- data.table(gene_name = tmp7$gene_name,
                 `Log2 Fold Change` = round(tmp7$log2FoldChange, 2),
                 `SE of Log2 Fold Change` = round(tmp7$log2FoldChangeSE, 2),
                 `Adjusted p-Value` = round(tmp7$padj, 3))

t4 <- merge(med,
            t4,
            by = "gene_name")

names(t4)[1:3] <- c("Gene",
                    "Median hits in non-COVID",
                    "Median hits in COVID")

write.csv(t4,
          file = "tmp/t4.csv",
          row.names = FALSE)
t4
```

```{r}
sum(t4$`Log2 Fold Change` > 0)
sum(t4$`Log2 Fold Change` < 0)
```

### Reactom PA

```{r}
gene_list_3 <- mapIds(org.Hs.eg.db,
                      keys = tmp7$gene_name,
                      column = "ENTREZID",
                      keytype = "SYMBOL")
length(gene_list_3)
```

```{r}
rpa3 <- ReactomePA::enrichPathway(gene = gene_list_3,
                                  readable = TRUE)

write.csv(rpa3@result,
          file = "tmp/Pathway Analysis of DM 431 differentially expresed genes.csv")

rpa3@result
```

## DM in COVID/non-COVID

### Prepare DDS

```{r}
ndx1 <- which(colnames(dt_rna_cod) %in% rnaseq_meta_clin$sample_names[rnaseq_meta_clin$AnyDM == 1])

tmp <- as.matrix(dt_rna_cod[,
                            ndx1,
                            with = FALSE])
dim(tmp)
rownames(tmp) <- dt_rna_cod$gene_id

dds_dm <- DESeqDataSetFromMatrix(countData = tmp, 
                                  colData = rnaseq_meta_clin[rnaseq_meta_clin$AnyDM == 1, ],
                                  design = ~ Cohort)
dds_dm
```

### Run DESeq

```{r}
deseq_cod_dm <- DESeq(dds_dm)
resultsNames(deseq_cod_dm)
```

### Results

```{r}
res <- results(deseq_cod_dm)
res[1:10, ]
rownames(res)[1:10]
```

```{r}
out <- data.table(gene_id = rownames(res),
                  baseMean = res$baseMean,
                  log2FoldChange = res$log2FoldChange,
                  log2FoldChangeSE = res$lfcSE,
                  stat = res$stat,
                  pvalue = res$pvalue,
                  padj = res$padj)

out <- merge(dt_rna_cod[, c("gene_name",
                            "gene_id")],
             out,
             by = "gene_id")

out <- out[!is.na(padj), ]

setorder(out, 
         padj)


head(out)

write.csv(out,
          file = "tmp/deseq2_cod_dm_only_covid_nocovid.CSV",
          row.names = FALSE)
```

```{r}
tmp8 <- out[abs(log2FoldChange) > 1 &
              padj < 0.05, ]
```

### Calculate the mean expressions of the selected genes

```{r}
tmp <- dt_rna_cod[gene_name %in% tmp8$gene_name, ]
tmp <- melt.data.table(data = tmp,
                       id.vars = 1,
                       measure.vars = 12:ncol(tmp),
                       variable.name = "sample_names",
                       value.name = "Hits")

tmp <- merge(tmp,
             rnaseq_meta_clin,
             by = "sample_names")
```

### Table 5: COVID/non-COVID, DM only

```{r}
med <- tmp[, .(med = median(Hits)),
           by = list(gene_name,
                     Cohort)]
med <- dcast.data.table(data = med,
                        gene_name ~ Cohort,
                        value.var = "med")
med

t5 <- data.table(gene_name = tmp8$gene_name,
                 `Log2 Fold Change` = round(tmp8$log2FoldChange, 2),
                 `SE of Log2 Fold Change` = round(tmp8$log2FoldChangeSE, 2),
                 `Adjusted p-Value` = round(tmp5$padj, 3))

t5 <- merge(med,
            t5,
            by = "gene_name")

names(t5)[1:3] <- c("Gene",
                    "Median hits in non-COVID",
                    "Median hits in COVID")

write.csv(t5,
          file = "tmp/t5.csv",
          row.names = FALSE)
t5
```

### Reactom Pathway Analysis

##### Map gene names to Entrez IDs

```{r}
dt_rpa8 <- merge(dt_rna[!is.na(entrez_id), 
                        c("gene_name",
                          "entrez_id")],
                 tmp8[, c("gene_name",
                          "log2FoldChange")])
```

##### Run Reactom PA

```{r}
rpa8 <- ReactomePA::enrichPathway(gene = dt_rpa8$entrez_id,
                                  readable = TRUE)

write.csv(rpa8@result,
          file = "tmp/Pathway Analysis of COVID vs non-COVID differentially expresed genes in DM.csv")

rpa8@result
```

# !CONTINUE HERE 2/22/24!

### DE in COVID and DM

```{r}
covid_and_dm <- t4$Gene[t4$Gene %in% t3$Gene]
covid_and_dm
```

```{r}
dt_f3 <- dt_rna_cod[gene_name %in% covid_and_dm, ]

dt_f3 <- melt.data.table(data = dt_f3,
                         id.vars = 1,
                         measure.vars = 11:ncol(dt_f3),
                         variable.name = "sample_names",
                         value.name = "Hits")
dt_f3 <- merge(dt_f3,
               rnaseq_meta_clin,
               by = "sample_names")
```

```{r,fig.height=10,fig.width=6}
p4 <- ggplot(dt_f3,
             aes(x = Cohort,
                 y = log2(Hits + 1))) +
  facet_wrap(~ gene_name,
             scale = "free_y",
             ncol = 1) +
  geom_boxplot(width = 0.3,
               outlier.shape = "") +
  geom_point(aes(fill = AnyDM,
                 group = sample_names),
             size = 2,
             shape = 21,
             alpha = 0.7,
             position = position_dodge(0.3)) +
  scale_x_discrete("COVID",
                   breaks = c("No_COVID",
                              "COVID"),
                   labels = c("No",
                              "Yes")) +
  scale_y_continuous("Number of hits",
                     breaks = seq(from = 1,
                                  to = 20,
                                  by = 2),
                     labels = (2^seq(from = 1,
                                     to = 20,
                                     by = 2)) - 1) +
  scale_fill_discrete("Any DM",
                      breaks = c(0, 1),
                      labels = c("No",
                                 "Yes")) +
  theme_bw() +
  theme(legend.position = "bottom")

p5 <- ggplot(dt_f3,
             aes(x = AnyDM,
                 y = log2(Hits + 1))) +
  facet_wrap(~ gene_name,
             scale = "free_y",
             ncol = 1) +
  geom_boxplot(width = 0.3,
               outlier.shape = "") +
  geom_point(aes(fill = Cohort,
                 group = sample_names),
             size = 2,
             shape = 21,
             alpha = 0.7,
             position = position_dodge(0.3)) +
  scale_x_discrete("Any DM",
                   breaks = c(0, 1),
                   labels = c("No",
                              "Yes")) +
  scale_y_continuous("Number of hits",
                     breaks = seq(from = 1,
                                  to = 20,
                                  by = 2),
                     labels = (2^seq(from = 1,
                                     to = 20,
                                     by = 2)) - 1) +
  scale_fill_discrete("COVID",
                      breaks = c("No_COVID",
                                 "COVID"),
                      labels = c("No",
                                 "Yes")) +
  theme_bw() +
  theme(legend.position = "bottom")

tiff(filename = "tmp/fig3_covid_dm.tiff",
     height = 10,
     width = 6,
     units = 'in',
     res = 600,
     compression = "lzw+p")
grid.arrange(p4,
             p5,
             nrow = 1)
graphics.off()

grid.arrange(p4,
             p5,
             nrow = 1)
```

### DESeq: AnyDM/no DM in COVID patients only

```{r}
samples_covid_only <- rnaseq_meta_clin$sample_names[rnaseq_meta_clin$Cohort == "COVID"]
ndx_covid_only <- which(colnames(dt_rna_cod) %in% samples_covid_only)

tmp <- as.matrix(dt_rna_cod[, ndx_covid_only, with = FALSE])
rownames(tmp) <- dt_rna_cod$gene_id

dds_cod_dm_covid_only <- DESeqDataSetFromMatrix(countData = tmp, 
                                                colData = rnaseq_meta_clin[sample_names %in% samples_covid_only],
                                                ~ AnyDM)
```

#### Filter out genes with small number of hits

```{r}
keep <- rowSums(counts(dds_cod_dm_covid_only) >= 10) >= 10
sum(keep)

dds_cod_dm_covid_only <- dds_cod_dm_covid_only[keep, ]
```

#### Run DESeq

```{r}
deseq_cod_dm_covid_only <- DESeq(dds_cod_dm_covid_only)

resultsNames(deseq_cod_dm_covid_only)
```

#### DESeq results

```{r}
res_dm_covid_only <- results(deseq_cod_dm_covid_only)
res_dm_covid_only[1:10, ]
rownames(res_dm_covid_only)[1:10]
```

```{r}
out_dm_covid_only <- data.table(gene_id = rownames(res_dm_covid_only),
                                baseMean = res_dm_covid_only$baseMean,
                                log2FoldChange = res_dm_covid_only$log2FoldChange,
                                log2FoldChangeSE = res_dm_covid_only$lfcSE,
                                stat = res_dm_covid_only$stat,
                                pvalue = res_dm_covid_only$pvalue,
                                padj = res_dm_covid_only$padj)

out_dm_covid_only <- merge(dt_rna_cod[, c("gene_name",
                                          "gene_id")],
                           out_dm_covid_only,
                           by = "gene_id")

out_dm_covid_only <- out_dm_covid_only[!is.na(padj), ]

setorder(out_dm_covid_only, 
         padj)


head(out_dm_covid_only)

write.csv(out_dm_covid_only,
          file = "tmp/deseq2_cod_dm_covid_only.CSV",
          row.names = FALSE)
```

```{r}
tmp7 <- out_dm_covid_only[abs(log2FoldChange) > 1 &
                            padj < 0.05, ]
```

#### Calculate the mean expressions of the selected genes

```{r}
tmp <- dt_rna_cod[gene_name %in% tmp7$gene_name, ]
tmp <- melt.data.table(data = tmp,
                       id.vars = 1,
                       measure.vars = 11:ncol(tmp),
                       variable.name = "sample_names",
                       value.name = "Hits")

tmp <- merge(tmp,
             rnaseq_meta_clin,
             by = "sample_names")
```

#### Table 3: DM/no DM significant DEs

```{r}
# med <- tmp[, .(med = median(Hits)),
#            by = list(gene_name,
#                      Cohort)]
# med <- dcast.data.table(data = med,
#                         gene_name ~ Cohort,
#                         value.var = "med")
# med
# 
# t4 <- data.table(gene_name = tmp6$gene_name,
#                  `Log2 Fold Change` = round(tmp6$log2FoldChange, 2),
#                  `SE of Log2 Fold Change` = round(tmp6$log2FoldChangeSE, 2),
#                  `Adjusted p-Value` = round(tmp6$padj, 3))
# 
# t4 <- merge(med,
#             t4,
#             by = "gene_name")
# 
# names(t4)[1:3] <- c("Gene",
#                     "Median hits in non-COVID",
#                     "Median hits in COVID")
# 
# write.csv(t4,
#           file = "tmp/t4.csv",
#           row.names = FALSE)
# t4
```

```{r}
# sum(t4$`Log2 Fold Change` > 0)
# sum(t4$`Log2 Fold Change` < 0)
```

# Session

```{r}
sessionInfo()
```
